[ews-build.webkit.org] Add concept of validators

JonWBedard opened the pull request:
#### d4b703ca9a5662d04ac431e58cede01e420ee25f
<pre>
[ews-build.webkit.org] Add concept of validators
<a href="https://bugs.webkit.org/show_bug.cgi?id=247708">https://bugs.webkit.org/show_bug.cgi?id=247708</a>
rdar://102168279

Reviewed by Ryan Haddad.

* Tools/CISupport/ews-build/steps.py:
(ValidateCommitterAndReviewer): Add list of validators for apple remote.
(ValidateCommitterAndReviewer.fail_build_due_to_invalid_status): Split from fail_build.
(ValidateCommitterAndReviewer.fail_build_due_to_no_validators):
(ValidateCommitterAndReviewer.fail_build): Split into fail_build and fail_invalid_status.
(ValidateCommitterAndReviewer.start): If a remote has validators, ensure that one of the
validators is in the list of reviewers.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/256543@main">https://commits.webkit.org/256543@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5910c64fdf8efb056014b8ad9d03e071d668e76

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105645 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100072 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5463 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34106 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88473 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101463 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101755 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82700 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39835 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/95058 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37510 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/20670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42106 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2175 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43996 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/39932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1701,7 +1704,7 @@ def getResultSummary(self):
         return buildstep.BuildStep.getResultSummary(self)
 
-    def fail_build(self, email_or_username, status):
+    def fail_invalid_status(self, email_or_username, status):
```
Nit: `fail_invalid_status` => `fail_build_due_to_invalid_status`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1754,17 +1761,36 @@ def start(self):
             reviewer = self.getProperty('reviewer', '').lower()
             reviewers = [reviewer] if reviewer else []
+
+        remote = self.getProperty('remote', DEFAULT_REMOTE)
```
would be good to separate it out as a method.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1754,17 +1761,36 @@ def start(self):
+                validator_list = f'{user_prefix}{validators[0]}'
+            else:
+                validator_list = f'{", ".join(f"{user_prefix}{v}" for v in validators[:-1])} or {user_prefix}{validators[-1]}'
+            reason = f"Landing changes on '{remote}' requires validation from {validator_list}"
```
Nit: `Landing changes on apple requires validation` might be a confusing error message for users. Might want to make it little more explicit, atleast say on `apple remote` instead of `apple`.

Also, adding a new term `validation`/`validator` might be confusing. How does someone validates? Isn't it just approval from specific users. Maybe the error message should just say `requires approval from ...` instead of `requires validation from ...`.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1691,6 +1691,9 @@ def send_email_for_github_failure(self):
     name = 'validate-commiter-and-reviewer'
     descriptionDone = ['Validated commiter and reviewer']
+    VALIDATORS_FOR = {
+        'apple': ['geoffreygaren', 'markcgee'],  # FIXME: This should be a bot, for now validation is manual
```
Do we want to hard-code the list here, or in `emails.json`?
Either way is fine with me, just wanted to point out an alternative.

Also, if this is going to be permanent, we can consider adding this info to `contributors.json` so that we don't have to maintain such lists at multiple places.

JonWBedard replied:
Discussed this offline with @aj062.

We _do_ want this hard-coded, it's going to be a bot in the near future, which will make this hard-coding more natural than it is now.

aj062 commented on Tools/CISupport/ews-build/steps_unittest.py:
```diff
@@ -5278,6 +5278,30 @@ def test_load_contributors_from_disk(self):
         contributors = filter(lambda element: element.get('name') == 'Aakash Jain', Contributors().load_from_disk()[0])
         self.assertEqual(list(contributors)[0]['emails'][0], 'aakash_jain@apple.com')
 
+    def test_success_pr_validators(self):
```
maybe we should add another unit-test explicitly for non-apple remote.

JonWBedard committed:
[ews-build.webkit.org] Add concept of validators
https://bugs.webkit.org/show_bug.cgi?id=247708
rdar://102168279

Reviewed by Ryan Haddad.

* Tools/CISupport/ews-build/steps.py:
(ValidateCommitterAndReviewer): Add list of validators for apple remote.
(ValidateCommitterAndReviewer.fail_build_due_to_invalid_status): Split from fail_build.
(ValidateCommitterAndReviewer.fail_build_due_to_no_validators):
(ValidateCommitterAndReviewer.fail_build): Split into fail_build and fail_invalid_status.
(ValidateCommitterAndReviewer.start): If a remote has validators, ensure that one of the
validators is in the list of reviewers.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/256543@main
